src/SegmentedString.chpl

```diff
@@ -1520,11 +1528,48 @@ module SegmentedString {
     return interpretAsString(values, rng, borrow=true).isAlpha();
   }
 
+  /*
+    The SegFunction called by computeOnSegments for isdecimal, using isDigit
+  */
+  inline proc stringIsDecimal(ref values, rng) throws {
+    return interpretAsString(values, rng, borrow=true).isDigit();
+  }
+
   /*
     The SegFunction called by computeOnSegments for isdigit
   */
   inline proc stringIsDigit(ref values, rng) throws {
-    return interpretAsString(values, rng, borrow=true).isDigit();
+    use In1d;
+    const specialDigits = "⁰¹²³⁴⁵⁶⁷⁸⁹₀₁₂₃₄₅₆₇₈₉";
+    const myString = interpretAsString(values, rng, borrow=true);
+    // test if string is all regular digits for early out
+    if myString.isDigit(){
+      return true;
+    }
+    // test if string is all alphanumeric characters for early out
+    else if myString.isAlpha() || myString.isEmpty() {
+      return false;
+    }
+    // string contains at least one special digit character, full test
+    else {
+      // this function converts bytes to a uint64 to keep all bytes together for the comparison
+      // by shifting by 8 to combine into a single uint64 to use In1d on integer values
+      proc toUint64(c) {
+        var ret:uint = 0;
+        var i = 0;
+        for b in c.bytes() {
+          var tmp = b:uint;
+          ret |= (tmp << (i*8));
+          i += 1;
+        }
+        return ret;
+      }
+      var specialArray = [s in specialDigits] toUint64(s);
+      var myStringArray = [b in myString] toUint64(b);
+      var digitTruth = [b in myString] b.isDigit();
```

Comment:
in re: to gasnet failing to compile. I was able to get it to compile (i haven't checked if it's working correctly) by making these distributed arrays.
```chapel
var specialArray = makeDistArray([s in specialDigits] toUint64(s));
var myStringArray = makeDistArray([b in myString] toUint64(b));
var digitTruth = makeDistArray([b in myString] b.isDigit());
```

So that seems to work. But it's prob not what we want to do because why distribute these small arrays when we don't need to. The issue is calling `uniqueSortWithInverse` with non-distributed arrays has 2 code paths, the normal one will return non-distributed arrays if that's what's passed in. But there's an early out is `size == 0` which is always distributed. So now there are cases where it sometimes returns distributed arrays and sometimes not and the compiler is rightfully upset about that lol

I'm looking into this more to see if i can find a better solution

Reply:
yah this is better, just change all the `makeDistArray` calls in `uniqueSortWithInverse ` to use `aD.size` and it fixes the issue. Diff included below
```diff
diff --git a/src/Unique.chpl b/src/Unique.chpl
index 56295452a..8dd14078e 100644
--- a/src/Unique.chpl
+++ b/src/Unique.chpl
@@ -62,9 +62,9 @@ module Unique
     proc uniqueSortWithInverse(a: [?aD] ?eltType) throws {
         if (aD.size == 0) {
             try! uLogger.debug(getModuleName(),getRoutineName(),getLineNumber(),"zero size");
-            var u = makeDistArray(0, eltType);
-            var c = makeDistArray(0, int);
-            var inv = makeDistArray(0, int);
+            var u = makeDistArray(aD.size, eltType);
+            var c = makeDistArray(aD.size, int);
+            var inv = makeDistArray(aD.size, int);
             return (u, c, inv);
         }
         var sorted = makeDistArray(aD, eltType);
@@ -84,7 +84,7 @@ module Unique
         // check there's enough room to create a copy for scan and throw if creating a copy would go over memory limit
         overMemLimit(numBytes(int) * bcast.size);
         bcast = (+ scan bcast);
-        var inv = makeDistArray(aD, int);
+        var inv = makeDistArray(aD.size, int);
         forall (p, b) in zip(perm, bcast) with (var agg = newDstAggregator(int)) {
             agg.copy(inv[p], b);
         }
```

Reply:
i just got off a video chat with jake, but thought it might be helpful for others to know how to go about debugging an issue like this. What I did was look at the locations of the error
```
/home/amandapotts/git/arkouda//src/Unique.chpl:62: error: unable to resolve return type
```
and saw line 62 of `Unique.chpl` was `uniqueSortWithInverse`. The error message says chpl is unable to figure out the return type, so I looked for all the spots where we return in that `proc` and tried to get their return types to be as similar as possible. `uniqueSortWithInverse` is included below but for simplicity I remove everything that's not a variable declaration or a return, because all that we care about
```chpl
proc uniqueSortWithInverse(a: [?aD] ?eltType) throws {
    if (aD.size == 0) {
        var u = makeDistArray(0, eltType);
        var c = makeDistArray(0, int);
        var inv = makeDistArray(0, int);
        return (u, c, inv);
    }
    var (u, c) = uniqueFromSorted(sorted);
    ...
    var inv = makeDistArray(aD, int);
    ...
    return (u, c, inv);
}
```

Looking at `uniqueFromSorted`, I see one code path is:
```chpl
proc uniqueFromSorted(sorted: [?aD] ?eltType, param needCounts = true) throws {
    ...
    if (allUnique == aD.size) {
        try!  uLogger.debug(getModuleName(),getRoutineName(),getLineNumber(),
                                "early out already unique");
        var u = makeDistArray(aD.size, eltType);
        u = sorted; // array is already unique
        if (needCounts) {
            var c = makeDistArray(aD.size, int);
            c = 1;
            return (u,c);
        } else {
            return u;
        }
    }
    ...
}
```
so `u` and `c` are declared with `makeDistArray(aD.size, int);`. so our two code paths are
```chpl
proc uniqueSortWithInverse(a: [?aD] ?eltType) throws {
    if (aD.size == 0) {
        var u = makeDistArray(0, eltType);
        var c = makeDistArray(0, int);
        var inv = makeDistArray(0, int);
        return (u, c, inv);
    }
    var u = makeDistArray(aD.size, eltType);
    var c = makeDistArray(aD.size, int);
    var inv = makeDistArray(aD, int);
    return (u, c, inv);
}
```
So if I change them all to `makeDistArray(aD.size, int)`, we'd be sure the return types line up. This doesn't mess up the early out because we know `aD.size == 0`

Reply:
Updated and confirmed that it builds and runs on gasnet